src/library/equations_compiler/elim_match.cpp

```diff
@@ -502,17 +503,14 @@ struct elim_match_fn {
                 expr const & p = head(eqn.m_patterns);
                 if (is_local(p)) {
                     has_variable = true; return true;
-                } else if (is_constructor_app(p)) {
+                }
+                type_context ctx = mk_type_context(eqn.m_lctx);
+                bool is_val = is_value(ctx, p);
+                if ((is_constructor_app(p) && (value_is_contructor || !is_val)) ||
+                    (value_is_contructor && is_val)) {
                     has_constructor = true; return true;
-                } else {
-                    type_context ctx = mk_type_context(eqn.m_lctx);
-                    if (is_value(ctx, p)) {
-                        if (value_is_contructor) has_constructor = true;
-                        return true;
-                    } else {
-                        return false;
-                    }
                 }
+                return false;
             });
         return r && has_variable && has_constructor;
     }
```

Comment:
Do you need this change for finishing this PR?
If not, let's discuss it after the paper deadline. 
When we use "value" transition, the equation lemmas do not hold definitionally.
One option is to *never* use value transition for 'enum' inductive datatypes.

Here are other pending issues regarding the equation compiler:
- We need to support nested inductive datatypes in the complete transition.
- Equation validation will miss invalid equations that require complete transition, and use inaccessible annotations.